### PR TITLE
Fixes podperson hair colors, adds individual layer colors

### PR DESCRIPTION
**Pull request by Time-Green**
This was broken completely by a certain scaly coder

Now it works again! Previously the hair was the inverted color of the body, but it should've been the same and only have the front layer be inverted

Also makes it possible to render layers with different colors

image

:cl:
fix: Podperson hair isnt fugly anymore
code: External organs now support different colors for different layers
/:cl:

### CODE REVIEW

**Comment by Time-Green** [comment]
It’s supposed to fail I just forgot to update the tests

File: code/modules/surgery/organs/external/_external_organs.dm [hunk 1]
```diff
@@ -156,6 +160,9 @@
 	if(sprite_datum.color_src)
 		appearance.color = draw_color
 
+	if(apply_color_to_layer)
+		appearance.color = layer_to_color_index["[image_layer]"]
+
```
Comment (LemonInTheDark) [inline code/modules/surgery/organs/external/_external_organs.dm:165]: shouldn't this code just be replaced by a get_overlay_color

File: code/modules/surgery/organs/external/_external_organs.dm [hunk 2]
```diff
@@ -250,13 +257,25 @@
...
-	color = draw_color
+	//Fuck giving the whole thing the same color, instead every layer can have its own
+	if(apply_color_to_layer)
+		layer_to_color_index = list()
+		for(var/draw_layer in all_layers)
+			if(draw_layer & layers)
+				//We make our layer a string because it'll runtime if we index with integers out of sequence
+				layer_to_color_index["[bitflag_to_layer(draw_layer)]"] = get_overlay_color(draw_layer)
+	else
+		color = draw_color
```
Comment (LemonInTheDark) [inline code/modules/surgery/organs/external/_external_organs.dm:268]: wait why does layer_to_color_index exist if get_overlay_color exists

File: code/modules/surgery/organs/external/_external_organs.dm [hunk 3]
```diff
@@ -424,6 +448,9 @@
...
 
-/obj/item/organ/external/pod_hair/override_color(rgb_value)
-	var/list/rgb_list = rgb2num(rgb_value)
-	return rgb(255 - rgb_list[1], 255 - rgb_list[2], 255 - rgb_list[3])
+/obj/item/organ/external/pod_hair/get_overlay_color(draw_layer)
+	if(draw_layer != color_swapped_layer)
+		return ..()
+
+	var/list/rgb_list = rgb2num(draw_color)
+	return rgb(color_inverse_base - rgb_list[1], color_inverse_base - rgb_list[2], color_inverse_base - rgb_list[3])
```
Comment (LemonInTheDark) [inline code/modules/surgery/organs/external/_external_organs.dm:456]: is it to cache this?

File: code/modules/surgery/organs/external/_external_organs.dm [hunk 4]
```diff
@@ -41,6 +41,10 @@
 	var/draw_color
 	///Where does this organ inherit it's color from?
 	var/color_source = ORGAN_COLOR_INHERIT
+	///FALSE will set the color on the organ, TRUE will call get_overlay_color() for every layer and put it directly on the overlay
+	var/apply_color_to_layer = FALSE
+	///Indexes our generated colors for each layer, only used if apply_color_to_layer is TRUE
+	var/list/layer_to_color_index
```
Comment (LemonInTheDark) [inline code/modules/surgery/organs/external/_external_organs.dm:47]: having these smells wrong, would prefer relying on just proc overrides if possible

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/surgery/organs/external/_external_organs.dm (hunk 2 above)
Comment (Time-Green) [reply code/modules/surgery/organs/external/_external_organs.dm:268]: The idea was to cache them, both so we controle when they change and to help with performance. Although I can just generate them with an overridable proc using just the draw_color and layer

**Comment by Time-Green** [comment]
I uncached the pod layer-colors

File: code/modules/surgery/organs/external/_external_organs.dm [hunk 5]
```diff
@@ -250,13 +255,18 @@
 				return
 			var/mob/living/carbon/human/human_owner = ownerlimb.owner
 			draw_color = human_owner.hair_color
-	color = draw_color
+	if(!apply_color_to_layer)
+		color = draw_color
```
Comment (LemonInTheDark) [inline code/modules/surgery/organs/external/_external_organs.dm:259]: couldn't we just make this 
```suggestion
	color = get_overlay_color(image_layer)
```

Comment (Time-Green) [reply code/modules/surgery/organs/external/_external_organs.dm:259]: The color here is the color that applies to the whole object. The actual overlays are generated in generate_and_fetch_overlays based on a bunch of mob characteristics and can’t be changed from here. Inherit colors isn’t called very frequently and is just to set the initial color to draw with

File: code/modules/surgery/organs/external/_external_organs.dm [hunk 6]
```diff
@@ -424,6 +439,9 @@
...
 
-/obj/item/organ/external/pod_hair/override_color(rgb_value)
-	var/list/rgb_list = rgb2num(rgb_value)
-	return rgb(255 - rgb_list[1], 255 - rgb_list[2], 255 - rgb_list[3])
+/obj/item/organ/external/pod_hair/get_overlay_color(draw_layer)
+	if(draw_layer != bitflag_to_layer(color_swapped_layer))
+		return ..()
+
+	var/list/rgb_list = rgb2num(draw_color)
+	return rgb(color_inverse_base - rgb_list[1], color_inverse_base - rgb_list[2], color_inverse_base - rgb_list[3])
```
[the remark this answers is not in the document]
Comment (Time-Green) [reply code/modules/surgery/organs/external/_external_organs.dm:447]: Eh I changed my mind, it’s super minor. It’s already cached in the limb_cache whatever so it’s just unneeded complexity

**Comment by Time-Green** [comment]
> image
> 
> you can fix this while you're here

????????

image
image

**Pull request merged**